### PR TITLE
[nn] TripletMarginLoss and PairwiseDistance : no batch dim

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #60585 

cc @albanD @mruberry @jbschlosser @walterddr

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -1436,9 +1436,9 @@ class TripletMarginLoss(_Loss):
             specifying either of those two args will override :attr:`reduction`. Default: ``'mean'``
 
     Shape:
-        - Input: :math:`(N, D)` where :math:`D` is the vector dimension.
-        - Output: A Tensor of shape :math:`(N)` if :attr:`reduction` is ``'none'``, or a scalar
-          otherwise.
+        - Input: :math:`(N, D)` or :math`(D)` where :math:`D` is the vector dimension.
+        - Output: A Tensor of shape :math:`(N)` if :attr:`reduction` is ``'none'`` and
+                  input shape is not :math`(D)`, or a scalar otherwise.
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:1441]: I think it reads slightly better to say `A Tensor of shape (N) if reduction is 'none' and input shape is (N, D); a scalar otherwise.`

File: torch/testing/_internal/common_nn.py [hunk 2]
```diff
@@ -5437,6 +5437,8 @@ def flatten(xs):
     ('SoftMarginLoss', lambda: torch.randn(9), lambda: torch.tensor([-1, 1, 1] * 3)),
     ('NLLLoss', lambda: F.log_softmax(torch.randn(3), dim=0), lambda: torch.tensor(1)),
     ('CosineEmbeddingLoss', lambda: (torch.randn(9), torch.randn(9)), lambda: torch.tensor(1)),
+    # For TripletMargingLoss, input_fn : (anchor, positive) and target_fn : negative
```
Comment (jbschlosser) [inline torch/testing/_internal/common_nn.py:5440]: typo: TripletMargingLoss -> TripletMarginLoss

File: aten/src/ATen/native/Loss.cpp [hunk 3]
```diff
@@ -79,10 +79,26 @@ Tensor hinge_embedding_loss(const Tensor& self, const Tensor& target, double mar
...
+      a_dim,
+      "D, ",
+      p_dim,
+      "D and ",
+      n_dim,
+      "D inputs.")
+  auto pairwise_distance = [p, eps](const Tensor& x1, const Tensor& x2) {
+    auto innermost_dim = x1.dim() - 1;
+    return at::norm(x1 - x2 + eps, p, innermost_dim, false);
+  };
```
Comment (jbschlosser) [inline aten/src/ATen/native/Loss.cpp:97]: Hm, I'm now rethinking my original decision to leave out `nn.PairwiseDistance` from supporting no batch dim. There's nothing that prevents it from being updated as you did here to support 1D inputs - that would simplify the logic here.

I updated the [rollup issue](https://github.com/pytorch/pytorch/issues/60585) to move `nn.PairwiseDistance` into the list of things to be updated.

**Review by jbschlosser — COMMENTED** [review]
Looking pretty good! Seeing this PR made me reexamine the idea of updating `at::pairwise_distance` / `nn.PairwiseDistance`. I now think the right thing to do there is to update those to support no batch dim, simplifying this PR - wdyt?

File: aten/src/ATen/native/Loss.cpp (hunk 3 above)
Comment (kshitij12345) [reply aten/src/ATen/native/Loss.cpp]: Done. Thanks :)!

**Comment by kshitij12345** [comment]
@jbschlosser should be ready for another round :) Thanks!

File: torch/testing/_internal/common_nn.py [hunk 4]
```diff
@@ -3935,6 +3943,23 @@ def single_batch_reference_fn(input, parameters, module):
         skip_double=TEST_WITH_ROCM,
         pickle=False,
     ),
+    dict(
+        module_name='PairwiseDistance',
+        input_fn=lambda: (torch.randn(10, 8), torch.randn(10, 8)),
+    ),
```
Comment (jbschlosser) [inline torch/testing/_internal/common_nn.py:3949]: Was PairwiseDistance untested in these dicts before? If so, might be worth adding a case or two for broadcasting as well

File: torch/nn/modules/distance.py [hunk 5]
```diff
@@ -18,8 +18,8 @@ class PairwiseDistance(Module):
         keepdim (bool, optional): Determines whether or not to keep the vector dimension.
             Default: False
     Shape:
-        - Input1: :math:`(N, D)` where `D = vector dimension`
-        - Input2: :math:`(N, D)`, same shape as the Input1
+        - Input1: :math:`(N, D)` or :math:`(D)` where `D = vector dimension` and `N = batch dimension`
+        - Input2: :math:`(N, D)` or :math:`(D)`, same shape as the Input1
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/nn/modules/distance.py:22]: How do you think it should be phrased?

```
- Output: :math:`(N)` or :math:`()` as per the Input shape. If :attr:`keepdim` is ``True``, then :math:`(N, 1) or :math:`(1)`.
```

Sorry I always get flustered with phrasing these things.
Thanks!

File: torch/testing/_internal/common_nn.py (hunk 4 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_nn.py:3955]: Done :)

File: torch/nn/modules/distance.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nn/modules/distance.py:22]: Hm, I think the way you phrased it is good! Just one minor thing, second shape in the first part should be `()` instead of `(D)`:
```
Output: :math:`(N)` or :math:`()` based on input dimension.
```

[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/nn/modules/distance.py:22]: Ah. Right 😅 Thanks!

**Review by jbschlosser — APPROVED** [review]
LGTM!
